File: typescript/packages/core/src/workspace/snapshot/state.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { CacheEntry } from '../../cache/file/entry.ts'
import { RAMFileCacheStore } from '../../cache/file/ram.ts'
import type { Resource } from '../../resource/base.ts'
import { EVENT_CLEAR, EVENT_COMMAND, EVENT_DELETE } from '../../observe/log_entry.ts'
import type { EventDict } from '../../observe/observer.ts'
import { RAMResource, type RAMResourceState } from '../../resource/ram/ram.ts'
import { z } from 'zod'

import { BIN_PREFIX } from '../../resource/bin/bin.ts'
import type { CLIInstall } from '../cli/types.ts'
import { HISTORY_PREFIX } from '../../resource/history/history.ts'
import {
  hasRedactedSecret,
  redactConfigWithSchema,
  resourceStateRequiresOverride,
} from '../../resource/secrets.ts'
import { Job, JobStatus } from '../../shell/job_table.ts'
import { ConsistencyPolicy, MountMode } from '../../types.ts'
import { VERSION } from '../../version.ts'
import type { NodeMeta } from '../mount/namespace/namespace.ts'
import { Session } from '../session/session.ts'
import type { Workspace } from '../workspace.ts'
import type { MountArgs } from './config.ts'
import { captureFingerprints, liveOnlyMountPrefixes } from './drift.ts'
import type {
  CacheEntrySnapshot,
  CLISnapshot,
  FingerprintEntrySnapshot,
  JobSnapshot,
  MountSnapshot,
  NodeMetaSnapshot,
  ResourceState,
  SessionSnapshot,
  WorkspaceStateDict,
} from './types.ts'
import { FORMAT_VERSION, normMountPrefix } from './utils.ts'

const VALID_MODES: readonly string[] = [MountMode.READ, MountMode.WRITE, MountMode.EXEC]

export async function toStateDict(ws: Workspace): Promise<WorkspaceStateDict> {
  const skip = new Set(['/dev/', normMountPrefix(HISTORY_PREFIX), normMountPrefix(BIN_PREFIX)])
  const mounts = [...ws.registry.allMounts()].filter((m) => !skip.has(m.prefix))
  const mountSnapshots: MountSnapshot[] = []
  for (let i = 0; i < mounts.length; i++) {
    const m = mounts[i]
    if (m === undefined) continue
    const resource = m.resource as unknown as {
      kind: string
      getState: () => ResourceState | Promise<ResourceState>
    }
    const state = await Promise.resolve(resource.getState())
    mountSnapshots.push({
      index: i,
      prefix: m.prefix,
      mode: m.mode,
      consistency: ConsistencyPolicy.LAZY,
      resource_class: resource.kind,
      resource_state: state,
    })
  }
  const ramCache = ws.cache instanceof RAMFileCacheStore ? ws.cache : null
  const cacheEntries: CacheEntrySnapshot[] =
    ramCache !== null
      ? ramCache.snapshotEntries().map(({ key, entry }) => ({
          key,
          data: ramCache.store.files.get(key) ?? new Uint8Array(),
          fingerprint: entry.fingerprint,
          ttl: entry.ttl,
          cached_at: entry.cachedAt,
          size: entry.size,
        }))
      : []
  const sessions: SessionSnapshot[] = ws.sessionManager
    .list()
    .map((s) => s.toJSON() as unknown as SessionSnapshot)
  const jobs: JobSnapshot[] = ws.jobTable
    .listJobs()
    .filter((j) => j.status !== JobStatus.RUNNING)
    .map((j) => ({
      id: j.id,
      command: j.command,
      cwd: j.cwd,
      status: j.status,
      stdout: j.stdout,
      stderr: j.stderr,
      exit_code: j.exitCode,
      created_at: j.createdAt,
      agent: j.agent,
      session_id: j.sessionId,
    }))
  const clisState: CLISnapshot[] = [...ws.registry.clis.items()].map(([name, install]) => ({
    name,
    spec: install.spec.name,
    config: captureCliConfig(install),
  }))
  const historyEvents = (await ws.observer.events()).filter(
    (e) => e.type === EVENT_COMMAND || e.type === EVENT_CLEAR || e.type === EVENT_DELETE,
  )
  const fingerprints: FingerprintEntrySnapshot[] = captureFingerprints(ws.records, ws.registry)
  const liveOnly = liveOnlyMountPrefixes(ws.registry)
  const nodes: Record<string, NodeMetaSnapshot> = {}
  for (const [path, meta] of ws.namespace.nodes) {
    const entry: NodeMetaSnapshot = {}
    if (meta.target !== undefined) entry.target = meta.target
    if (meta.mtime !== undefined) entry.mtime = meta.mtime
    if (meta.mode !== undefined) entry.mode = meta.mode
    if (meta.uid !== undefined) entry.uid = meta.uid
    if (meta.gid !== undefined) entry.gid = meta.gid
    if (meta.atime !== undefined) entry.atime = meta.atime
    nodes[path] = entry
  }
  return {
    version: FORMAT_VERSION,
    mirage_version: VERSION,
    default_session_id: ws.sessionManager.defaultId,
    default_agent_id: ws.agentId,
    current_agent_id: ws.agentId,
    sessions,
    mounts: mountSnapshots,
    cache: {
      limit: ws.cache.cacheLimit,
      max_drain_bytes: ramCache !== null ? ramCache.maxDrainBytes : null,
      entries: cacheEntries,
    },
    history: historyEvents,
    jobs,
    fingerprints,
    live_only_mounts: liveOnly,
    nodes,
    clis: clisState,
  }
}

function captureCliConfig(install: CLIInstall): Record<string, unknown> | null {
  const model = install.spec.configModel
  if (model instanceof z.ZodObject) {
    return redactConfigWithSchema(model, install.config)
  }
  if (install.config !== null && typeof install.config === 'object') {
    return install.config as Record<string, unknown>
  }
  return null
}

export function buildMountArgs(
  state: WorkspaceStateDict,
  overrides: Record<string, Resource> = {},
  cliOverrides: Record<string, Record<string, unknown>> = {},
): MountArgs {
  if (state.version < FORMAT_VERSION) {
    throw new Error(
      `snapshot format v${String(state.version)} not supported ` +
        `(loader expects v${String(FORMAT_VERSION)})`,
    )
  }
  const normalized: Record<string, Resource> = {}
  for (const [prefix, resource] of Object.entries(overrides)) {
    normalized[normMountPrefix(prefix)] = resource
  }
  const missing = state.mounts
    .filter(
      (m) =>
        normalized[normMountPrefix(m.prefix)] === undefined &&
        resourceStateRequiresOverride(m.resource_state),
    )
    .map((m) => m.prefix)
  if (missing.length > 0) {
    throw new Error(
      `Workspace.load: resources= must include overrides for: ${missing.join(', ')}. ` +
        `These mounts were saved with redacted creds or transient connection state ` +
        `and need fresh resources.`,
    )
  }
  const mountArgs: Record<string, [Resource, MountMode]> = {}
  for (const m of state.mounts) {
    if (!VALID_MODES.includes(m.mode)) {
      throw new Error(`Workspace.fromState: mount '${m.prefix}' has invalid mode '${m.mode}'`)
    }
    mountArgs[m.prefix] = [
      normalized[normMountPrefix(m.prefix)] ?? new RAMResource(),
      m.mode as MountMode,
    ]
  }
  const cliEntries = state.clis ?? []
  const missingClis = cliEntries
    .filter((e) => hasRedactedSecret(e.config) && !(e.name in cliOverrides))
    .map((e) => e.name)
  if (missingClis.length > 0) {
    throw new Error(
      `Workspace.load: clis= must include fresh configs for: ${missingClis.join(', ')}. ` +
        `These CLIs were saved with redacted config secrets.`,
    )
  }
  const cliArgs: Record<string, [string, Record<string, unknown> | null]> = {}
  for (const e of cliEntries) {
    cliArgs[e.name] = [e.spec, cliOverrides[e.name] ?? e.config]
  }

  return {
    mountArgs,
    consistency: ConsistencyPolicy.LAZY,
    defaultSessionId: state.default_session_id,
    defaultAgentId: state.default_agent_id,
    ...(cliEntries.length > 0 ? { clis: cliArgs } : {}),
  }
}

export async function applyStateDict(ws: Workspace, state: WorkspaceStateDict): Promise<void> {
  for (const m of state.mounts) {
    if (resourceStateRequiresOverride(m.resource_state)) continue
    const mount = ws.registry.mountFor(m.prefix)
    if (mount === null) continue
    const resource = mount.resource as unknown as {
      loadState: (state: ResourceState) => void | Promise<void>
    }
    await Promise.resolve(resource.loadState(m.resource_state as RAMResourceState))
  }
  await restoreSessions(ws, state)
  // current_agent_id is not restored separately: TS models a single
  // readonly agentId, set to default_agent_id at construction (== current).
  restoreCache(ws, state)
  await restoreHistory(ws, state)
  restoreJobs(ws, state)
  await restoreNodes(ws, state)
}

async function restoreNodes(ws: Workspace, state: WorkspaceStateDict): Promise<void> {
  const entries = new Map<string, NodeMeta>()
  for (const [path, e] of Object.entries(state.nodes ?? {})) {
    const meta: NodeMeta = {}
    if (e.target !== undefined) meta.target = e.target
    if (e.mtime !== undefined) meta.mtime = e.mtime
    if (e.mode !== undefined) meta.mode = e.mode
    if (e.uid !== undefined) meta.uid = e.uid
    if (e.gid !== undefined) meta.gid = e.gid
    if (e.atime !== undefined) meta.atime = e.atime
    entries.set(path, meta)
  }
  await ws.namespace.replaceNodes(entries)
}

async function restoreSessions(ws: Workspace, state: WorkspaceStateDict): Promise<void> {
  // The snapshot's default session identity wins over the live one,
  // and the discovery record's pointer follows it. A state without the
  // pointer (older commit metas) keeps the live default, mirroring the
  // Python None-guard.
  if (state.default_session_id != null) {
    await ws.adoptDefaultSession(state.default_session_id)
  }
  const restored: Session[] = []
  for (const s of state.sessions) {
    const exists = ws.sessionManager.list().some((x) => x.sessionId === s.session_id)
    const session = exists
      ? ws.sessionManager.get(s.session_id)
      : ws.sessionManager.create(s.session_id)
    const fields = Session.fromJSON(s)
    session.cwd = fields.cwd
    session.env = fields.env
    session.mountModes = fields.mountModes
    restored.push(session)
  }
  // The snapshot's session table wins over prior store contents,
  // mirroring Namespace.replaceNodes.
  await ws.sessionManager.replaceFromSnapshot(restored)
}

function restoreCache(ws: Workspace, state: WorkspaceStateDict): void {
  if (!(ws.cache instanceof RAMFileCacheStore)) return
  for (const e of state.cache.entries) {
    ws.cache.loadEntry(
      e.key,
      e.data,
      new CacheEntry({
        size: e.size,
        cachedAt: e.cached_at,
        fingerprint: e.fingerprint,
        ttl: e.ttl,
      }),
    )
  }
}

async function restoreHistory(ws: Workspace, state: WorkspaceStateDict): Promise<void> {
  // Always load (loadEvents clears first): a snapshot with empty history
  // still rewinds the recorder, same as the cache clear. Foreign-format
  // entries (e.g. a Python snapshot's different history shape) are skipped
  // inside loadEvents.
  await ws.observer.loadEvents((state.history as EventDict[] | undefined) ?? [])
}

function restoreJobs(ws: Workspace, state: WorkspaceStateDict): void {
  for (const j of state.jobs) {
    ws.jobTable.loadJob(
      new Job({
        id: j.id,
        command: j.command,
        cwd: j.cwd,
        agent: j.agent,
        sessionId: j.session_id,
        createdAt: j.created_at,
        status: j.status as JobStatus,
        stdout: j.stdout,
        stderr: j.stderr,
        exitCode: j.exit_code,
      }),
    )
  }
}

File: python/mirage/workspace/snapshot/state.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import importlib
import tempfile
from typing import Any

from mirage.observe.log_entry import EVENT_CLEAR, EVENT_COMMAND, EVENT_DELETE
from mirage.resource.bin import BIN_PREFIX
from mirage.resource.history import HISTORY_PREFIX
from mirage.resource.registry import REGISTRY, resolve_class
from mirage.resource.secrets import (has_redacted_secret, redacted_config_dump,
                                     revealed_config_dump)
from mirage.shell.job_table import Job, JobStatus
from mirage.types import (CacheKey, CLIKey, ConsistencyPolicy, JobKey,
                          MountKey, MountMode, ResourceName, ResourceStateKey,
                          SessionKey, StateKey)
from mirage.version import __version__
from mirage.workspace.mount.namespace import NodeMeta
from mirage.workspace.session.session import Session
from mirage.workspace.snapshot.config import MountArgs
from mirage.workspace.snapshot.drift import (capture_fingerprints,
                                             live_only_mount_prefixes)
from mirage.workspace.snapshot.utils import FORMAT_VERSION, norm_mount_prefix


async def to_state_dict(ws) -> dict[str, Any]:
    auto_prefixes = {
        "/dev/",
        norm_mount_prefix(HISTORY_PREFIX),
        norm_mount_prefix(BIN_PREFIX),
    }

    mounts_state = []
    for idx, m in enumerate(mt for mt in ws._registry.mounts()
                            if mt.prefix not in auto_prefixes):
        mounts_state.append({
            MountKey.INDEX: idx,
            MountKey.PREFIX: m.prefix,
            MountKey.MODE: m.mode.value,
            MountKey.CONSISTENCY: m.consistency.value,
            MountKey.RESOURCE_CLASS:
            f"{type(m.resource).__module__}.{type(m.resource).__name__}",
            MountKey.RESOURCE_STATE: m.resource.get_state(),
        })

    cache = ws._cache
    cache_entries = [{
        CacheKey.KEY: k,
        CacheKey.DATA: cache._store.files.get(k, b""),
        CacheKey.FINGERPRINT: e.fingerprint,
        CacheKey.TTL: e.ttl,
        CacheKey.CACHED_AT: e.cached_at,
        CacheKey.SIZE: e.size,
    } for k, e in cache._entries.items()]

    history_events = [
        e for e in await ws.observer.events()
        if e.get("type") in (EVENT_COMMAND, EVENT_CLEAR, EVENT_DELETE)
    ]

    clis_state = [{
        CLIKey.NAME:
        name,
        CLIKey.SPEC:
        install.spec.name,
        CLIKey.CONFIG: (redacted_config_dump(install.config)
                        if install.config is not None else None),
    } for name, install in ws._registry.clis.items().items()]

    finished_jobs = [
        _job_to_dict(j) for j in ws.job_table.list_jobs()
        if j.status != JobStatus.RUNNING
    ]

    fingerprints = capture_fingerprints(ws)
    live_only_mounts = live_only_mount_prefixes(ws)

    return {
        StateKey.VERSION: FORMAT_VERSION,
        StateKey.MIRAGE_VERSION: __version__,
        StateKey.MOUNTS: mounts_state,
        StateKey.SESSIONS: [s.to_dict() for s in ws._session_mgr.list()],
        StateKey.DEFAULT_SESSION_ID: ws._session_mgr.default_id,
        StateKey.DEFAULT_AGENT_ID: ws._default_agent_id,
        StateKey.CURRENT_AGENT_ID: ws._current_agent_id,
        StateKey.CACHE: {
            CacheKey.LIMIT: cache.cache_limit,
            CacheKey.MAX_DRAIN_BYTES: cache.max_drain_bytes,
            CacheKey.ENTRIES: cache_entries,
        },
        StateKey.HISTORY: history_events,
        StateKey.CLIS: clis_state,
        StateKey.JOBS: finished_jobs,
        StateKey.FINGERPRINTS: fingerprints,
        StateKey.LIVE_ONLY_MOUNTS: live_only_mounts,
        StateKey.NODES: {
            path: meta.to_fields()
            for path, meta in ws._namespace.nodes.items()
        },
    }


def build_mount_args(
        state: dict[str, Any],
        resources: dict[str, Any] | None = None,
        clis: dict[str, dict[str, Any]] | None = None) -> MountArgs:
    """Translate a state dict into Workspace constructor inputs.

    Validates that every mount with redacted secrets has a resource
    override, and every CLI installed with a redacted config has a
    fresh config override.
    Does NOT construct a Workspace — that's the caller's job.

    Raises:
        ValueError: if any redacted mount or CLI lacks an override, or
            if the snapshot is from an unsupported format version.
    """
    saved_version = state.get(StateKey.VERSION)
    if saved_version is not None and saved_version < FORMAT_VERSION:
        raise ValueError(f"snapshot format v{saved_version} not supported "
                         f"(loader expects v{FORMAT_VERSION}); "
                         "regenerate via `mirage workspace snapshot`")

    overrides = {norm_mount_prefix(k): v for k, v in (resources or {}).items()}

    missing = [
        m[MountKey.PREFIX] for m in state[StateKey.MOUNTS]
        if requires_resource_override(m)
        and norm_mount_prefix(m[MountKey.PREFIX]) not in overrides
    ]
    if missing:
        raise ValueError(
            "Workspace.load: resources= must include overrides for: "
            f"{missing}. These mounts were saved with redacted creds "
            "or transient connection state and need fresh resources.")

    cli_overrides = clis or {}
    cli_entries = state.get(StateKey.CLIS) or []
    missing_clis = [
        e[CLIKey.NAME] for e in cli_entries
        if has_redacted_secret(e[CLIKey.CONFIG])
        and e[CLIKey.NAME] not in cli_overrides
    ]
    if missing_clis:
        raise ValueError(
            "Workspace.load: clis= must include fresh configs for: "
            f"{missing_clis}. These CLIs were saved with redacted "
            "config secrets.")

    mount_args: dict[str, tuple[Any, ...]] = {}
    for m in state[StateKey.MOUNTS]:
        prefix = norm_mount_prefix(m[MountKey.PREFIX])
        prov = (overrides[prefix]
                if prefix in overrides else _construct_resource(m))
        mount_args[m[MountKey.PREFIX]] = (prov, MountMode(m[MountKey.MODE]))

    cli_args = {
        e[CLIKey.NAME]:
        (e[CLIKey.SPEC], cli_overrides.get(e[CLIKey.NAME], e[CLIKey.CONFIG]))
        for e in cli_entries
    }

    return MountArgs(
        mount_args=mount_args,
        consistency=ConsistencyPolicy.LAZY,
        default_session_id=state[StateKey.DEFAULT_SESSION_ID],
        default_agent_id=state.get(StateKey.DEFAULT_AGENT_ID),
        clis=cli_args or None,
    )


async def apply_state_dict(ws, state: dict[str, Any]) -> None:
    """Restore post-construction state into an already-built Workspace.

    Restores: resource load_state (content, fresh disk root, etc.),
    sessions, cache entries, history, finished jobs.

    Workspace must already have its mounts constructed via the args
    from build_mount_args. This function is purely additive — it does
    not construct anything.
    """
    # load_state runs for ALL mounts (overridden too), so disk content
    # is written into the new root, redis content into the new URL, etc.
    # Cred-only resources (S3 et al.) define load_state as no-op.
    for m in state[StateKey.MOUNTS]:
        try:
            mount = ws._registry.mount_for_prefix(m[MountKey.PREFIX])
        except ValueError:
            continue
        mount.resource.load_state(m[MountKey.RESOURCE_STATE])

    await _restore_sessions(ws, state)
    ws._current_agent_id = state.get(StateKey.CURRENT_AGENT_ID,
                                     ws._default_agent_id)

    _restore_cache(ws, state)
    await _restore_history(ws, state)
    _restore_jobs(ws, state)
    await _restore_nodes(ws, state)


async def _restore_nodes(ws, state: dict[str, Any]) -> None:
    entries = {
        path: NodeMeta.from_fields(d)
        for path, d in (state.get(StateKey.NODES) or {}).items()
    }
    await ws._namespace.replace_nodes(entries)


async def _restore_sessions(ws, state: dict[str, Any]) -> None:
    default_sid = state.get(StateKey.DEFAULT_SESSION_ID)
    if default_sid is not None:
        # The snapshot's default session identity wins over the live
        # one, and the discovery record's pointer follows it.
        ws._session_mgr.adopt_default(default_sid)
        ws._default_session_id = default_sid
        await ws._state_store.replace_meta(ws._workspace_id, {
            "workspace_id": ws._workspace_id,
            "default_session_id": default_sid,
        })
        ws._meta_written = True
    restored: list[Any] = []
    for s_data in state.get(StateKey.SESSIONS, []):
        sid = s_data[SessionKey.SESSION_ID]
        if sid == default_sid:
            session = ws._session_mgr.get(sid)
        else:
            try:
                session = ws._session_mgr.create(sid)
            except ValueError:
                # The session already exists live (checkout on a
                # running workspace): the restored state wins, matching
                # the replace_from_snapshot contract below.
                session = ws._session_mgr.get(sid)
        fields = Session.from_dict(s_data)
        session.cwd = fields.cwd
        session.env = fields.env
        session.mount_modes = fields.mount_modes
        restored.append(session)
    # The snapshot's session table wins over prior store contents,
    # mirroring Namespace.replace_nodes.
    await ws._session_mgr.replace_from_snapshot(restored)


def _restore_cache(ws, state: dict[str, Any]) -> None:
    cache_state = state.get(StateKey.CACHE) or {}
    if hasattr(ws._cache, "max_drain_bytes"):
        ws._cache.max_drain_bytes = cache_state.get(CacheKey.MAX_DRAIN_BYTES)
    cache = ws._cache
    if not hasattr(cache, "_entries") or not hasattr(cache, "_store"):
        # Non-RAM cache backend (e.g. Redis) — skip; its content lives
        # outside the workspace and isn't part of the snapshot anyway.
        return
    from mirage.cache.file.entry import CacheEntry
    for entry in cache_state.get(CacheKey.ENTRIES, []):
        key = entry[CacheKey.KEY]
        data = entry[CacheKey.DATA]
        cache._store.files[key] = data
        cache._entries[key] = CacheEntry(
            size=entry.get(CacheKey.SIZE, len(data)),
            cached_at=entry.get(CacheKey.CACHED_AT, 0),
            fingerprint=entry.get(CacheKey.FINGERPRINT),
            ttl=entry.get(CacheKey.TTL),
        )
        cache._cache_size += entry.get(CacheKey.SIZE, len(data))


async def _restore_history(ws, state: dict[str, Any]) -> None:
    # Always load (load_events clears first): a snapshot with empty
    # history still rewinds the recorder, same as the cache clear.
    await ws.observer.load_events(state.get(StateKey.HISTORY) or [])


def _restore_jobs(ws, state: dict[str, Any]) -> None:
    max_id = 0
    for job_d in state.get(StateKey.JOBS, []):
        max_id = max(max_id, job_d.get(JobKey.ID, 0))
        ws.job_table._jobs[job_d[JobKey.ID]] = _job_from_dict(job_d)
    ws.job_table._next_id = max_id + 1


def _job_to_dict(job) -> dict[str, Any]:
    return {
        JobKey.ID: job.id,
        JobKey.COMMAND: job.command,
        JobKey.CWD: job.cwd,
        JobKey.STATUS: job.status.value,
        JobKey.STDOUT: job.stdout,
        JobKey.STDERR: job.stderr,
        JobKey.EXIT_CODE: job.exit_code,
        JobKey.CREATED_AT: job.created_at,
        JobKey.AGENT: job.agent,
        JobKey.SESSION_ID: job.session_id,
    }


def _job_from_dict(d: dict[str, Any]):
    return Job(
        id=d[JobKey.ID],
        command=d[JobKey.COMMAND],
        task=None,
        cwd=d.get(JobKey.CWD, "/"),
        status=JobStatus(d.get(JobKey.STATUS, JobStatus.COMPLETED.value)),
        stdout=d.get(JobKey.STDOUT, b"") or b"",
        stderr=d.get(JobKey.STDERR, b"") or b"",
        exit_code=d.get(JobKey.EXIT_CODE, 0),
        created_at=d.get(JobKey.CREATED_AT, 0.0),
        agent=d.get(JobKey.AGENT, "unknown"),
        session_id=d.get(JobKey.SESSION_ID, "default"),
    )


def _construct_resource(mount_state: dict[str, Any]):
    cls = _resource_class_for(mount_state)
    resource_state = mount_state[MountKey.RESOURCE_STATE]
    ptype = resource_state.get(ResourceStateKey.TYPE, "")

    if ptype == ResourceName.RAM:
        return cls()
    if ptype == ResourceName.DISK:
        return cls(root=tempfile.mkdtemp(prefix="mirage-disk-"))
    if ptype == ResourceName.REDIS:
        raise ValueError(
            f"Redis mount at {mount_state[MountKey.PREFIX]} requires "
            "resources= override")

    config = resource_state.get(ResourceStateKey.CONFIG)
    if config is None:
        return cls()
    config_cls = _config_class_for(cls)
    if config_cls is not None:
        return cls(config_cls(**config))
    return cls()


def requires_resource_override(mount_state: dict[str, Any]) -> bool:
    resource_state = mount_state[MountKey.RESOURCE_STATE]
    config = resource_state.get(ResourceStateKey.CONFIG)
    config_cls = _config_class_for(_resource_class_for(mount_state))
    return has_redacted_secret(config, config_cls)


def reusable_clis(ws, state: dict[str, Any]) -> dict[str, dict[str, Any]]:
    """Fresh-config overrides a copy needs for secret-bearing CLIs.

    A CLI saved with a redacted config cannot reinstall from the state
    dict alone; a same-process copy still holds the live validated
    config, so its secrets are revealed back into an override the way
    remote mounts share their live resources.

    Args:
        ws: the origin workspace.
        state (dict[str, Any]): the origin's state dict.
    """
    live = ws._registry.clis.items()
    overrides: dict[str, dict[str, Any]] = {}
    for e in state.get(StateKey.CLIS) or []:
        name = e[CLIKey.NAME]
        install = live.get(name)
        if (has_redacted_secret(e[CLIKey.CONFIG]) and install is not None
                and install.config is not None):
            overrides[name] = revealed_config_dump(install.config)
    return overrides


def reusable_resources(mounts: list[Any], state: dict[str,
                                                      Any]) -> dict[str, Any]:
    """Live resources a copy should share with its origin.

    Remote backends (S3, Redis, GDrive) stay shared: their state
    redacts the secrets a reconstruction would need. Local content
    resources (RAM, Disk) are rebuilt fresh so the copy's writes do
    not clobber the original's data. The auto mounts are excluded
    because the new workspace mounts its own.

    Args:
        mounts (list[Any]): the origin's mount entries.
        state (dict[str, Any]): the origin's state dict.
    """
    auto = {
        "/dev/",
        norm_mount_prefix(HISTORY_PREFIX),
        norm_mount_prefix(BIN_PREFIX),
    }
    live = {m.prefix: m.resource for m in mounts if m.prefix not in auto}
    return {
        m[MountKey.PREFIX]: live[m[MountKey.PREFIX]]
        for m in state[StateKey.MOUNTS]
        if requires_resource_override(m) and m[MountKey.PREFIX] in live
    }


def _resource_class_for(mount_state: dict[str, Any]):
    ptype = mount_state[MountKey.RESOURCE_STATE].get(ResourceStateKey.TYPE, "")
    if ptype in REGISTRY:
        return resolve_class(REGISTRY[ptype].resource_path)
    cls_path = mount_state[MountKey.RESOURCE_CLASS]
    mod_name, cls_name = cls_path.rsplit(".", 1)
    return getattr(importlib.import_module(mod_name), cls_name)


def _config_class_for(resource_cls):
    mod = importlib.import_module(resource_cls.__module__)
    for name in dir(mod):
        obj = getattr(mod, name)
        if isinstance(obj, type) and name.endswith("Config"):
            return obj
    return None
